scripts/python/proposers_rewards.py

```diff
@@ -0,0 +1,416 @@
+from typing import Dict, Optional, List, Tuple, Generator, TextIO, Union
+
+import sys
+import asyncio
+import json
+import itertools
+import enum
+import os
+from collections import namedtuple
+
+import pydantic
+import httpx
+import typer
+
+# VIT servicing station models
+
+ADA = '₳'
+FUNDED = "FUNDED"
+NOT_FUNDED = "NOT_FUNDED"
+YES = "YES"
+NO = "NO"
+NOT_FUNDED_OVER_BUDGET = "Not Funded - Over Budget"
+NOT_FUNDED_APPROVAL_THRESHOLD = "Not Funded - Approval Threshold"
+LOVELACE_FACTOR = 1000000
+
+
+class Proposal(pydantic.BaseModel):
+    internal_id: int
+    proposal_id: str
+    proposal_title: str
+    proposal_funds: int
+    proposal_url: str
+    chain_proposal_id: str
+    chain_proposal_index: int
+    chain_vote_options: Dict[str, int]
+    fund_id: int
+    challenge_id: int
+    challenge_type: str
+
+
+# Jormungandr models
+
+class Options(pydantic.BaseModel):
+    start: int
+    end: int
+
+
+class Results(pydantic.BaseModel):
+    results: List[int]
+
+
+# we mimic the tally
+class TallyResult(pydantic.BaseModel):
+    result: Results
+
+
+class DecryptedTally(pydantic.BaseModel):
+    decrypted: TallyResult
+
+
+class PrivateTallyState(pydantic.BaseModel):
+    state: DecryptedTally
+
+
+class PrivateTally(pydantic.BaseModel):
+    Private: PrivateTallyState
+
+    @property
+    def results(self):
+        try:
+            return self.Private.state.decrypted.result.results
+        except AttributeError:
+            return None
+
+
+class PublicTally(pydantic.BaseModel):
+    Public: TallyResult
+
+    @property
+    def results(self):
+        try:
+            return self.Public.result.results
+        except AttributeError:
+            return None
+
+
+class ProposalStatus(pydantic.BaseModel):
+    index: int
+    proposal_id: str
+    options: Options
+    tally: Optional[Union[PublicTally, PrivateTally]]
+    votes_cast: int
+
+
+class VoteplanStatus(pydantic.BaseModel):
+    id: str
+    payload: str
+    proposals: List[ProposalStatus]
+
+
+class Challenge(pydantic.BaseModel):
+    id: int
+    challenge_type: str
+    title: str
+    description: str
+    rewards_total: int
+    proposers_rewards: int
+    fund_id: int
+    challenge_url: str
+
+
+# File loaders
+
+def load_json_from_file(file_path: str) -> Dict:
+    with open(file_path, encoding="utf-8") as f:
+        return json.load(f)
+
+
+def get_proposals_from_file(proposals_file_path: str) -> Dict[str, Proposal]:
+    proposals: Generator[Proposal, None, None] = (
+        Proposal(**proposal_data) for proposal_data in load_json_from_file(proposals_file_path)
```

Comment:
what is the advantage of using a generator instead of list comprehension?

Reply:
It is not consumed immediately which avoids the intermediate allocation.

Reply:
ahh, so also the dict below is created piece by piece upon request?

Reply:
Yeah, generators are lazy so it is consumed by the dict below.